### PR TITLE
Fix MGTwoLevelTransfer when used without MPI

### LINKED ISSUE #15512: MGTwoLevelTransfer not working without MPI
When trying to call `MGTwoLevelTransfer` on a machine without MPI, we crash with the message:
```
An error occurred in line <109> of file </home/kronbichler/deal/other/deal.II/source/base/mpi_compute_index_owner_internal.cc> in function
    FlexibleIndexStorage::index_type &dealii::Utilities::MPI::internal::ComputeIndexOwner::FlexibleIndexStorage::operator[](const std::size_t)
The violated condition was: 
    ::dealii::deal_II_exceptions::internals::compare_less_than(index, size)
Additional information: 
    Index 0 is not in the half-open range [0,0). In the current case, this
    half-open range is in fact empty, suggesting that you are accessing an
    element of an empty collection such as a vector that has not been set
    to the correct size.
```

The problem is that we have code to set up `FlexibleIndexStorage` inside an `#ifdef DEAL_II_WITH_MPI`: https://github.com/dealii/dealii/blob/8d68b3192ed1c8b928dfce03ca1d615201859904/source/base/mpi_compute_index_owner_internal.cc#L167-L189 Unless we make MPI a required dependency, I think this is very unfortunate. I suggest to go through all functions and be much more conservative in which places to guard with `DEAL_II_WITH_MPI`.

As a related note, also when setting up deal.II with MPI but forgetting to call `MPI_Init()`, we get plenty of errors because we call `MPI_Comm_rank` (via `this_mpi_process`). Since adding `job_supports_mpi()` to all use cases is tedious, I suggest to also add `job_supports_mpi()` to the `MPI_Comm_rank` and `MPI_Comm_size` commands.

### PR DESCRIPTION
**Pull request by kronbichler**
This fixes #15512 by adding the necessary checks to enable the two-level transfer both for the case when deal.II was compiled without support for MPI (i.e., no `DEAL_II_WITH_MPI`) or when `MPI_Init` was not called for a serial program.

Built on top of #15516, so the other PR must be merged first. I will rebase once that is merged to more easily see the relevant changes. The second commit holds the actual change in the MG transfer in case this is reviewed already now. I consider this an important fix because the transfer is severely broken in serial.

### CODE REVIEW

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 1]
```diff
@@ -607,17 +600,27 @@ namespace internal
...
+      const unsigned int my_rank =
+        Utilities::MPI::this_mpi_process(communicator);
 
       {
         std::vector<types::global_dof_index> indices;
 
         for (const auto &i : targets_with_indexset)
           {
+            if (i.first == my_rank)
+              continue;
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:622]: It looks like this change causes the code to hang. It means the code depends on having an MPI rank sending MPI messages to itself. Not the only problematic performance here, as observed in #14968. I will look into this later.

Comment (kronbichler) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:627]: OK, addressed, I simply had a wrong condition further down.

File: tests/multigrid-global-coarsening/mg_transfer_p_07.cc [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+/**
+ * Test transfer operator for polynomial coarsening.
```
Comment (peterrum) [inline tests/multigrid-global-coarsening/mg_transfer_p_07.cc:18]: Could you reference the base test which has been extended here?

File: tests/multigrid-global-coarsening/mg_transfer_a_04.cc [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+/**
+ * Test transfer operator for global coarsening.
```
Comment (peterrum) [inline tests/multigrid-global-coarsening/mg_transfer_a_04.cc:18]: Same here.

**Review by peterrum — APPROVED** [review]

**Comment by kronbichler** [comment]
Yes, adding the comments on where these tests came from makes sense. I also shortened the test output of these two tests to more manageable sizes, to be more in line with what I preach.

**Pull request merged**